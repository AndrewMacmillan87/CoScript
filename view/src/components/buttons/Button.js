import React from "react";
import PropTypes from "prop-types";

import "./buttons.css";

const Button = ({ onClick, id, classes, value }) => {
    return (
        <button onClick={onClick} id={id} className={classes} type='button'>
            <span className='btn-text'>
                <b>{value}</b>
            </span>
        </button>
    );
};

Button.propTypes = {
    onClick: PropTypes.func,
    classes: PropTypes.string,
    id: PropTypes.string,
    value: PropTypes.string
};

export default Button;
