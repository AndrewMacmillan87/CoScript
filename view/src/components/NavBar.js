import React, { Fragment } from "react";
import { Link, withRouter } from "react-router-dom";
import Button from "./buttons/Button";

import { getUserMessages } from "../redux/auxiliary/user-messages";

// Redux
import { connect } from "react-redux";
import { getMostRecentProject } from "../redux/actions/documents";

import PropTypes from "prop-types";

import "./navbar.css";

export class NavBar extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            visible: false,
            unread: false,
        };
        this.handleClick = this.handleClick.bind(this);
    }

    /**
     * Checks to see if the user has any unread messages
     * in their personal profile message inbox and sets
     * the 'unread' state property to true if they have.
     * This displays a small icon to the user to indicate
     * they have unread messages
     */
    async componentDidMount() {
        const userMessages = await getUserMessages();
        for (let message of userMessages) {
            if (!message.read) {
                this.setState({
                    unread: true,
                });
            }
        }
    }

    /**
     * Sets the 'visible' state property so that the
     * dropdown menu can be revealed and hidden by
     * clicking the link
     */
    handleClick() {
        this.setState((state) => ({
            visible: !state.visible,
        }));
    }

    render() {
        const { unread } = this.state;
        const { user, loading, history } = this.props;

        const menuIcon = require("../img/menu.svg");
        const spinner = require("../img/loading.svg");
        const isLoggedIn = user && user.isLoggedIn | false;

        const loggedIn = (
            <Fragment>
                <div className='header'>
                    <div className='logo-container'>
                        <Link to='/'>
                            <p className='logo'>CoScript</p>
                        </Link>
                    </div>
                    <div className='nav-links'>
                        <Link to='/projects'>Active Projects</Link>
                        <Link to='/publications'>Publications</Link>
                        <Link
                            to='#'
                            onClick={async () => {
                                await this.props.getMostRecentProject();
                                const { projectData } = this.props;
                                projectData
                                    ? history.push(`/editor${projectData.url}`)
                                    : history.push("/editor");
                            }}>
                            Workspace
                        </Link>
                    </div>
                </div>
                <div className='nav-btns'>
                    {user && !loading ? (
                        <div className='logged-in-user-container'>
                            <p className='user-username'>
                                {user.profile.displayName}
                            </p>
                            <button
                                className='desktop-icon-link'
                                onClick={this.handleClick}>
                                {unread ? (
                                    <div className='avatar-unread-msg-container'>
                                        <div className='avatar-unread-msg-icon'></div>
                                        <img
                                            className='user-gravatar'
                                            src={user.avatar}
                                            alt='user gravatar'
                                        />
                                    </div>
                                ) : (
                                    <img
                                        className='user-gravatar'
                                        src={user.avatar}
                                        alt='user gravatar'
                                    />
                                )}
                            </button>
                        </div>
                    ) : (
                        <img
                            className='user-spinner'
                            src={spinner}
                            alt='loading spinner icon'
                        />
                    )}
                </div>
                <button
                    className='dropdown-icon-link'
                    onClick={this.handleClick}>
                    <img
                        className='dropdown-icon'
                        src={menuIcon}
                        alt='Burger menu icon'
                    />
                </button>
                <div
                    id='dropdown-nav'
                    className={
                        this.state.visible
                            ? "dropdown-menu dropdown-content display"
                            : "dropdown-menu dropdown-content"
                    }>
                    <Link to='/projects'>
                        <p className='dropdown-link'>Active Projects</p>
                    </Link>
                    <Link to='/publications'>
                        <p className='dropdown-link'>Publications</p>
                    </Link>
                    <Link
                        to='#'
                        onClick={async () => {
                            await this.props.getMostRecentProject();
                            const { projectData } = this.props;
                            projectData
                                ? history.push(`/editor${projectData.url}`)
                                : history.push("/editor");
                        }}>
                        <p className='dropdown-link'>Workspace</p>
                    </Link>
                    <Link to='/profile'>
                        <p className='dropdown-link'>My Profile</p>
                    </Link>
                    <a
                        onClick={() => {
                            history["from"] = "logout";
                            history.push("/redirect");
                        }}
                        href=' '>
                        <p className='dropdown-link'>Logout</p>
                    </a>
                </div>
            </Fragment>
        );

        const visiting = (
            <Fragment>
                <div className='header'>
                    <div className='logo-container'>
                        <Link to='/'>
                            <p className='logo'>CoScript</p>
                        </Link>
                    </div>
                    <div className='nav-links'>
                        <Link to='/projects'>Active Projects</Link>
                        <Link to='/publications'>Publications</Link>
                    </div>
                </div>
                <div className='nav-btns'>
                    <div>
                        <Link to='/register'>
                            <Button classes='btn' value='Register' />
                        </Link>
                    </div>
                    <div>
                        <Link to='/login'>
                            <Button classes='grn-btn' value='Login' />
                        </Link>
                    </div>
                </div>
                <button
                    className='dropdown-icon-link'
                    onClick={this.handleClick}>
                    <img
                        className='dropdown-icon'
                        src={menuIcon}
                        alt='Burger menu icon'
                    />
                </button>
                <div
                    id='dropdown-nav'
                    className={
                        this.state.visible
                            ? "dropdown-menu dropdown-content display"
                            : "dropdown-menu dropdown-content"
                    }>
                    <Link to='/register'>
                        <p className='dropdown-link'>Register</p>
                    </Link>
                    <Link to='/login'>
                        <p className='dropdown-link'>Login</p>
                    </Link>
                    <Link to='/projects'>
                        <p className='dropdown-link'>Active Projects</p>
                    </Link>
                    <Link to='/publications'>
                        <p className='dropdown-link'>Publications</p>
                    </Link>
                </div>
            </Fragment>
        );

        return (
            <nav className='navbar'>
                {!loading && (
                    <Fragment>{!isLoggedIn ? visiting : loggedIn}</Fragment>
                )}
            </nav>
        );
    }
}

NavBar.propTypes = {
    user: PropTypes.object,
    loading: PropTypes.bool,
    getMostRecentProject: PropTypes.func.isRequired,
};

const setProps = (state) => ({
    user: state.auth.user,
    loading: state.auth.loading,
    projectData: state.documents.projectData,
});

export default connect(setProps, {
    getMostRecentProject,
})(withRouter(NavBar));
