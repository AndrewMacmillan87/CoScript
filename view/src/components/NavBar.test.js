import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router, Link } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../redux/store/reduxStore";

import { NavBar } from "./NavBar";
import Button from "./buttons/Button";

describe("NavBar component", () => {
    let props, mockGetMostRecentProject;
    beforeEach(() => {
        mockGetMostRecentProject = jest.fn();
        props = {
            user: {
                isLoggedIn: false,
                avatar: "",
                profile: { displayName: "Test Name" },
            },
            loading: false,
            getMostRecentProject: mockGetMostRecentProject,
            history: {
                from: "",
                push: jest.fn(),
            },
        };
    });

    test("renders without crashing", () => {
        global.scrollTo = jest.fn();
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <NavBar {...props} />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        global.scrollTo = jest.fn();
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <NavBar {...props} />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a nav element with a className of 'navbar'", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <NavBar {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("nav").length).toEqual(1);
        expect(wrapper.find(".navbar").length).toEqual(1);
    });

    describe("User not logged in", () => {
        it("should render 9 Link components", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <NavBar {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find(Link).length).toEqual(9);
        });

        it("should render 2 Button components", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <NavBar {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find(Button).length).toEqual(2);
        });

        it("should render 7 div components", () => {
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <NavBar {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("div").length).toEqual(7);
        });
    });

    describe("User is logged in", () => {
        it("should render 8 Link components", () => {
            props.user.isLoggedIn = true;
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <NavBar {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find(Link).length).toEqual(8);
        });

        it("should render 6 div elements", () => {
            props.user.isLoggedIn = true;
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <NavBar {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("div").length).toEqual(6);
        });

        it("should render 2 img elements", () => {
            props.user.isLoggedIn = true;
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <NavBar {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("img").length).toEqual(2);
        });

        it("should render 2 button elements", () => {
            props.user.isLoggedIn = true;
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <NavBar {...props} />
                    </Router>
                </Provider>
            );
            expect(wrapper.find("button").length).toEqual(2);
        });

        it("should render a div element with a className of 'avatar-unread-msg-icon' if user has unread messages", () => {
            props.user.isLoggedIn = true;
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <NavBar {...props} />
                    </Router>
                </Provider>
            );
            wrapper.find(NavBar).setState({ unread: true });
            expect(wrapper.find(".avatar-unread-msg-icon").length).toEqual(1);
        });

        it("should call `getMostRecentProject()` when 'Workspace' link is clicked", async () => {
            props.user.isLoggedIn = true;
            const wrapper = mount(
                <Provider store={reduxStore}>
                    <Router>
                        <NavBar {...props} />
                    </Router>
                </Provider>
            );
            wrapper.find(Link).at(6).simulate("click");
            await expect(mockGetMostRecentProject).toBeCalled();
        });
    });
});
