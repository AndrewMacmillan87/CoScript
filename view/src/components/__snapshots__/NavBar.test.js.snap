// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar component renders correctly 1`] = `
<nav
  className="navbar"
>
  <div
    className="header"
  >
    <div
      className="logo-container"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <p
          className="logo"
        >
          CoScript
        </p>
      </a>
    </div>
    <div
      className="nav-links"
    >
      <a
        href="/projects"
        onClick={[Function]}
      >
        Active Projects
      </a>
      <a
        href="/publications"
        onClick={[Function]}
      >
        Publications
      </a>
    </div>
  </div>
  <div
    className="nav-btns"
  >
    <div>
      <a
        href="/register"
        onClick={[Function]}
      >
        <button
          className="btn"
          type="button"
        >
          <span
            className="btn-text"
          >
            <b>
              Register
            </b>
          </span>
        </button>
      </a>
    </div>
    <div>
      <a
        href="/login"
        onClick={[Function]}
      >
        <button
          className="grn-btn"
          type="button"
        >
          <span
            className="btn-text"
          >
            <b>
              Login
            </b>
          </span>
        </button>
      </a>
    </div>
  </div>
  <button
    className="dropdown-icon-link"
    onClick={[Function]}
  >
    <img
      alt="Burger menu icon"
      className="dropdown-icon"
      src={
        Object {
          "ReactComponent": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "__esModule": true,
          "default": "menu.svg",
        }
      }
    />
  </button>
  <div
    className="dropdown-menu dropdown-content"
    id="dropdown-nav"
  >
    <a
      href="/register"
      onClick={[Function]}
    >
      <p
        className="dropdown-link"
      >
        Register
      </p>
    </a>
    <a
      href="/login"
      onClick={[Function]}
    >
      <p
        className="dropdown-link"
      >
        Login
      </p>
    </a>
    <a
      href="/projects"
      onClick={[Function]}
    >
      <p
        className="dropdown-link"
      >
        Active Projects
      </p>
    </a>
    <a
      href="/publications"
      onClick={[Function]}
    >
      <p
        className="dropdown-link"
      >
        Publications
      </p>
    </a>
  </div>
</nav>
`;
